src/imb/M3Publisher.cpp

```diff
@@ -0,0 +1,100 @@
+        }
+        else{
+            publisher_ = this->create_publisher<sensor_msgs::msg::PointCloud2>(this->get_parameter("topic").as_string().c_str(), 10);
+        }
+
+        data_vector_ = std::vector<uint8_t>();
+        packet_ready_ = false;
+        std::thread(&M3Publisher::CreateListener, this, this->get_parameter("ip_addr").as_string(), this->get_parameter("port").as_int()).detach();
+        timer_ = this->create_wall_timer(std::chrono::milliseconds(this->get_parameter("callback_time").as_int()), std::bind(&M3Publisher::ProcessData, this)); // How often to sample data from listener
+    }
+    void M3Publisher::ProcessData() {
+        try{
+            if(packet_ready_){
+                std::unique_lock<std::mutex> lock(mutex_); // Thread-safe handling
+                const uint8_t* start = &data_vector_[0]; // Grabs the pointer to the first element in the vector
+                // imb::ImbPacketStructure packet(start); // TODO: - fix packet structure
+
+                // Cast data to the correct types
+                imb::PacketHeader packet_header(start);
+                imb::DataHeader data_header(start + sizeof(imb::PacketHeader));
+                imb::DataBody data_body(start + sizeof(imb::PacketHeader) + sizeof(imb::DataHeader),
+                    data_header.nNumBeams, data_header.nNumImageSample, packet_header.dataType);
+
+                //Inintialize the pointcloud
+                pcl::PointCloud<pcl::PointXYZI> cloud;
+                cloud.height = 1;
+                cloud.width = data_header.nNumBeams * data_header.nNumImageSample;
+                cloud.is_dense = false;
+                cloud.points.resize(cloud.height * cloud.width);
+
+                for(size_t i = 0; i < data_header.nNumBeams; i++){
+                    float beam_angle = data_header.fBeamList[i]; //Current beam angle
+                    for(size_t j = 0; j < data_header.nNumImageSample; j++){
+                        pcl::PointXYZI point;
+                        float dist_to_point = ((data_header.fSWST - data_header.fTXWST) + data_header.fImageSampleInterval * j) 
+                            * data_header.fSoundSpeed / 2; // Formula given in the IMB format documentation
+                        point.x = dist_to_point * cos(beam_angle * M_PI / 180.0);
+                        point.y = dist_to_point * sin(beam_angle * M_PI / 180.0);
+                        point.z = 0;
+                        float intensity = abs(data_body.complexData(i, j));
+                        if(intensity > 0){
+                            point.intensity = intensity;
+                        }
+                        else{
+                            point.intensity = 0;
+                        }
+                        cloud.points[i * data_header.nNumImageSample + j] = point;
+                    }
+                }
```

Comment:
Not a fan of the math being calculated in the ros node class. Could this be moved to the `imb::DataBody` class perhaps?